src/gtc/gtir_to_oir.py

```diff
@@ -18,192 +18,177 @@
         )
         combined_mask = current_mask
         if mask:
             combined_mask = oir.BinaryOp(op=LogicalOperator.AND, left=mask, right=combined_mask)
-        self.visit(node.true_branch.body, mask=combined_mask, ctx=ctx)
+        stmts.extend(self.visit(node.true_branch.body, mask=combined_mask, ctx=ctx, **kwargs))
 
         if node.false_branch:
             combined_mask = oir.UnaryOp(op=UnaryOperator.NOT, expr=current_mask)
             if mask:
                 combined_mask = oir.BinaryOp(op=LogicalOperator.AND, left=mask, right=combined_mask)
-            self.visit(
-                node.false_branch.body,
-                mask=combined_mask,
-                ctx=ctx,
-            )
+            stmts.extend(self.visit(node.false_branch.body, mask=combined_mask, ctx=ctx, **kwargs))
+
+        return stmts
 
     # For now we represent ScalarIf (and FieldIf) both as masks on the HorizontalExecution.
     # This is not meant to be set in stone...
     def visit_ScalarIfStmt(
         self, node: gtir.ScalarIfStmt, *, mask: oir.Expr = None, ctx: Context, **kwargs: Any
-    ) -> None:
+    ) -> List[oir.Stmt]:
         current_mask = self.visit(node.cond)
         combined_mask = current_mask
         if mask:
             combined_mask = oir.BinaryOp(op=LogicalOperator.AND, left=mask, right=current_mask)
 
-        self.visit(node.true_branch.body, mask=combined_mask, ctx=ctx)
+        stmts = self.visit(node.true_branch.body, mask=combined_mask, ctx=ctx, **kwargs)
         if node.false_branch:
             combined_mask = oir.UnaryOp(op=UnaryOperator.NOT, expr=current_mask)
             if mask:
                 combined_mask = oir.BinaryOp(op=LogicalOperator.AND, left=mask, right=combined_mask)
-            self.visit(
-                node.false_branch.body,
-                mask=combined_mask,
-                ctx=ctx,
-            )
+            stmts.extend(self.visit(node.false_branch.body, mask=combined_mask, ctx=ctx, **kwargs))
+
+        return stmts
 
-    def visit_Interval(self, node: gtir.Interval, **kwargs: Any) -> oir.Interval:
+    # --- Misc ---
+    def visit_Interval(self, node: gtir.Interval) -> oir.Interval:
         return oir.Interval(
             start=self.visit(node.start),
             end=self.visit(node.end),
         )
 
-    def visit_VerticalLoop(
-        self, node: gtir.VerticalLoop, *, ctx: Context, **kwargs: Any
-    ) -> oir.VerticalLoop:
-        ctx.horizontal_executions.clear()
-        self.visit(node.body, ctx=ctx)
+    # --- Control flow ---
+    def visit_VerticalLoop(self, node: gtir.VerticalLoop, *, ctx: Context) -> oir.VerticalLoop:
+        horiz_execs: List[oir.HorizontalExecution] = []
+        for stmt in node.body:
+            ctx.reset_local_scalars()
+            ret = self.visit(stmt, ctx=ctx)
+            stmts = utils.flatten_list([ret] if isinstance(ret, oir.Stmt) else ret)
```

Comment:
I don't love this "hack", but otherwise I think the lowering is improved by this change.